File: nautilus_trader/adapters/alpaca/http.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------
from __future__ import annotations

import os
import urllib.parse
from typing import Any

from nautilus_trader.adapters.alpaca.constants import ALPACA_DATA_BASE_URL
from nautilus_trader.adapters.alpaca.constants import ALPACA_LIVE_TRADING_BASE_URL
from nautilus_trader.adapters.alpaca.constants import ALPACA_PAPER_TRADING_BASE_URL


try:
    import aiohttp
except ImportError as exc:  # pragma: no cover
    raise ImportError(
        "The Alpaca adapter requires aiohttp. Install with `nautilus_trader[alpaca]`.",
    ) from exc


class AlpacaHttpClient:
    def __init__(
        self,
        *,
        api_key: str | None,
        api_secret: str | None,
        paper: bool,
        trading_base_url: str | None = None,
        data_base_url: str | None = None,
        timeout_secs: int = 10,
    ) -> None:
        self.api_key = api_key or os.getenv("ALPACA_API_KEY") or os.getenv("APCA_API_KEY_ID")
        self.api_secret = (
            api_secret
            or os.getenv("ALPACA_API_SECRET")
            or os.getenv("APCA_API_SECRET_KEY")
        )
        if not self.api_key or not self.api_secret:
            raise ValueError(
                "Alpaca credentials not configured. Set config.api_key/config.api_secret "
                "or ALPACA_API_KEY/ALPACA_API_SECRET (or APCA_API_KEY_ID/APCA_API_SECRET_KEY).",
            )

        self.paper = paper
        self.trading_base_url = trading_base_url or (
            ALPACA_PAPER_TRADING_BASE_URL if paper else ALPACA_LIVE_TRADING_BASE_URL
        )
        self.data_base_url = data_base_url or ALPACA_DATA_BASE_URL
        self.timeout_secs = timeout_secs
        self._session: aiohttp.ClientSession | None = None

    @property
    def auth_headers(self) -> dict[str, str]:
        return {
            "APCA-API-KEY-ID": self.api_key,
            "APCA-API-SECRET-KEY": self.api_secret,
        }

    async def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            timeout = aiohttp.ClientTimeout(total=self.timeout_secs)
            self._session = aiohttp.ClientSession(timeout=timeout, headers=self.auth_headers)
        return self._session

    async def close(self) -> None:
        if self._session is not None and not self._session.closed:
            await self._session.close()

    async def _request(
        self,
        method: str,
        path: str,
        *,
        base: str,
        params: dict[str, Any] | None = None,
        payload: dict[str, Any] | None = None,
    ) -> Any:
        session = await self._get_session()
        base_url = self.trading_base_url if base == "trading" else self.data_base_url
        url = f"{base_url}{path}"
        params = {k: v for k, v in (params or {}).items() if v is not None}

        async with session.request(method, url, params=params, json=payload) as response:
            if response.status == 204:
                return None

            content_type = response.headers.get("Content-Type", "")
            if response.status >= 400:
                body = await response.json() if "application/json" in content_type else await response.text()
                raise RuntimeError(f"Alpaca request failed [{response.status}] {body}")

            if "application/json" in content_type:
                return await response.json()
            return await response.text()

    async def get_account(self) -> dict[str, Any]:
        return await self._request("GET", "/v2/account", base="trading")

    async def get_assets(
        self,
        *,
        status: str | None = None,
        asset_class: str | None = None,
    ) -> list[dict[str, Any]]:
        return await self._request(
            "GET",
            "/v2/assets",
            base="trading",
            params={"status": status, "asset_class": asset_class},
        )

    async def get_asset(self, symbol_or_asset_id: str) -> dict[str, Any]:
        encoded = urllib.parse.quote(symbol_or_asset_id, safe="")
        return await self._request("GET", f"/v2/assets/{encoded}", base="trading")

    async def get_option_contracts(
        self,
        *,
        underlying_symbols: list[str] | None = None,
        status: str | None = None,
        expiration_date_gte: str | None = None,
        expiration_date_lte: str | None = None,
        option_type: str | None = None,
        style: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        params = {
            "status": status,
            "expiration_date_gte": expiration_date_gte,
            "expiration_date_lte": expiration_date_lte,
            "type": option_type,
            "style": style,
            "limit": limit,
            "page_token": page_token,
        }
        if underlying_symbols:
            params["underlying_symbols"] = ",".join(underlying_symbols)

        return await self._request(
            "GET",
            "/v2/options/contracts",
            base="trading",
            params=params,
        )

    async def get_option_contract(self, symbol_or_contract_id: str) -> dict[str, Any]:
        encoded = urllib.parse.quote(symbol_or_contract_id, safe="")
        return await self._request(
            "GET",
            f"/v2/options/contracts/{encoded}",
            base="trading",
        )

    async def get_positions(self) -> list[dict[str, Any]]:
        return await self._request("GET", "/v2/positions", base="trading")

    async def list_orders(
        self,
        *,
        status: str = "all",
        limit: int | None = None,
        after: str | None = None,
        until: str | None = None,
        symbols: list[str] | None = None,
        nested: bool = False,
        direction: str | None = None,
        page_token: str | None = None,
    ) -> list[dict[str, Any]]:
        params = {
            "status": status,
            "limit": limit,
            "after": after,
            "until": until,
            "nested": str(nested).lower(),
            "direction": direction,
            "page_token": page_token,
        }
        if symbols:
            params["symbols"] = ",".join(symbols)
        return await self._request("GET", "/v2/orders", base="trading", params=params)

    async def get_order(self, order_id: str) -> dict[str, Any]:
        return await self._request("GET", f"/v2/orders/{order_id}", base="trading")

    async def get_order_by_client_order_id(self, client_order_id: str) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v2/orders:by_client_order_id",
            base="trading",
            params={"client_order_id": client_order_id},
        )

    async def submit_order(self, payload: dict[str, Any]) -> dict[str, Any]:
        return await self._request("POST", "/v2/orders", base="trading", payload=payload)

    async def replace_order(self, order_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        return await self._request(
            "PATCH",
            f"/v2/orders/{order_id}",
            base="trading",
            payload=payload,
        )

    async def cancel_order(self, order_id: str) -> None:
        await self._request("DELETE", f"/v2/orders/{order_id}", base="trading")

    async def cancel_all_orders(self) -> list[dict[str, Any]]:
        response = await self._request("DELETE", "/v2/orders", base="trading")
        return response or []

    async def get_activities(
        self,
        *,
        activity_type: str,
        after: str | None = None,
        until: str | None = None,
        page_size: int | None = None,
        direction: str | None = None,
        page_token: str | None = None,
    ) -> list[dict[str, Any]]:
        return await self._request(
            "GET",
            f"/v2/account/activities/{activity_type}",
            base="trading",
            params={
                "after": after,
                "until": until,
                "page_size": page_size,
                "direction": direction,
                "page_token": page_token,
            },
        )

    async def get_stock_quotes(
        self,
        *,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v2/stocks/quotes",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_stock_trades(
        self,
        *,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v2/stocks/trades",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_stock_bars(
        self,
        *,
        symbols: list[str],
        timeframe: str,
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v2/stocks/bars",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "timeframe": timeframe,
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_option_quotes(
        self,
        *,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v1beta1/options/quotes",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_option_trades(
        self,
        *,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v1beta1/options/trades",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_option_bars(
        self,
        *,
        symbols: list[str],
        timeframe: str,
        start: str,
        end: str,
        limit: int | None,
        feed: str,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            "/v1beta1/options/bars",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "timeframe": timeframe,
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "feed": feed,
                "page_token": page_token,
            },
        )

    async def get_crypto_quotes(
        self,
        *,
        loc: str,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            f"/v1beta3/crypto/{loc}/quotes",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "page_token": page_token,
            },
        )

    async def get_crypto_trades(
        self,
        *,
        loc: str,
        symbols: list[str],
        start: str,
        end: str,
        limit: int | None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            f"/v1beta3/crypto/{loc}/trades",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "page_token": page_token,
            },
        )

    async def get_crypto_bars(
        self,
        *,
        loc: str,
        symbols: list[str],
        timeframe: str,
        start: str,
        end: str,
        limit: int | None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET",
            f"/v1beta3/crypto/{loc}/bars",
            base="data",
            params={
                "symbols": ",".join(symbols),
                "timeframe": timeframe,
                "start": start,
                "end": end,
                "limit": limit,
                "sort": "asc",
                "page_token": page_token,
            },
        )

File: crates/adapters/alpaca/tests/http.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Integration tests for the Alpaca HTTP client using a mock Axum server.

use std::sync::{
    Arc,
    atomic::{AtomicUsize, Ordering},
};

use axum::{
    Router,
    body::Bytes,
    extract::{Path, Query, State},
    http::{HeaderMap, StatusCode},
    response::{IntoResponse, Response},
    routing::{delete, get, patch, post},
};
use nautilus_alpaca::{
    common::{
        credential::Credential,
        enums::{
            AlpacaEnvironment, AlpacaOrderSide, AlpacaOrderStatus, AlpacaOrderType,
            AlpacaTimeInForce,
        },
    },
    http::{
        client::{AlpacaHttpClient, AlpacaRawHttpClient},
        error::AlpacaHttpError,
        query::{
            GetAssetsParamsBuilder, GetOrdersParamsBuilder, GetStockBarsParamsBuilder,
            PatchOrderParamsBuilder, PostOrderParamsBuilder,
        },
    },
};
use nautilus_model::instruments::Instrument;
use nautilus_network::retry::{RetryConfig, RetryManager};
use ustr::Ustr;

const HTTP_GET_ASSETS: &str = include_str!("../test_data/http_get_assets.json");
const HTTP_GET_ACCOUNT: &str = include_str!("../test_data/http_get_account.json");
const HTTP_GET_ORDER: &str = include_str!("../test_data/http_get_order.json");
const HTTP_GET_ORDERS: &str = include_str!("../test_data/http_get_orders.json");
const HTTP_GET_POSITIONS: &str = include_str!("../test_data/http_get_positions.json");
const HTTP_GET_STOCK_BARS: &str = include_str!("../test_data/http_get_stock_bars.json");
const HTTP_GET_STOCK_BARS_PAGE2: &str = include_str!("../test_data/http_get_stock_bars_page2.json");
const HTTP_GET_STOCK_TRADES: &str = include_str!("../test_data/http_get_stock_trades.json");
const HTTP_GET_STOCK_QUOTES: &str = include_str!("../test_data/http_get_stock_quotes.json");
const HTTP_ERROR: &str = include_str!("../test_data/http_error.json");

const TEST_API_KEY: &str = "PKTEST1234567890";
const TEST_API_SECRET: &str = "supersecretvalue";

async fn spawn_server(router: Router) -> String {
    let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
    let addr = listener.local_addr().unwrap();
    tokio::spawn(async move { axum::serve(listener, router).await.unwrap() });
    format!("http://{addr}")
}

// Compressed retry timings keep retry-path assertions in the millisecond range.
fn fast_retry_manager(max_retries: u32) -> RetryManager<AlpacaHttpError> {
    RetryManager::new(RetryConfig {
        max_retries,
        initial_delay_ms: 1,
        max_delay_ms: 1,
        backoff_factor: 1.0,
        jitter_ms: 0,
        operation_timeout_ms: Some(60_000),
        immediate_first: true,
        max_elapsed_ms: Some(60_000),
    })
}

fn authenticated_raw_client(base_url: &str) -> AlpacaRawHttpClient {
    let credential = Credential::new(TEST_API_KEY.to_string(), TEST_API_SECRET.to_string());
    let mut client = AlpacaRawHttpClient::new(
        AlpacaEnvironment::Paper,
        Some(base_url.to_string()),
        Some(base_url.to_string()),
        10,
        None,
        Some(credential),
    )
    .unwrap();
    client.set_retry_manager(fast_retry_manager(3));
    client
}

fn authenticated_client(base_url: &str) -> AlpacaHttpClient {
    AlpacaHttpClient::from_raw(authenticated_raw_client(base_url))
}

fn assert_auth_headers(headers: &HeaderMap) {
    assert_eq!(
        headers.get("APCA-API-KEY-ID").and_then(|v| v.to_str().ok()),
        Some(TEST_API_KEY),
    );
    assert_eq!(
        headers
            .get("APCA-API-SECRET-KEY")
            .and_then(|v| v.to_str().ok()),
        Some(TEST_API_SECRET),
    );
}

async fn handle_account(headers: HeaderMap) -> Response {
    assert_auth_headers(&headers);
    ([("content-type", "application/json")], HTTP_GET_ACCOUNT).into_response()
}

async fn handle_assets(
    headers: HeaderMap,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Response {
    assert_auth_headers(&headers);
    assert_eq!(params.get("status").map(String::as_str), Some("active"));
    assert_eq!(
        params.get("asset_class").map(String::as_str),
        Some("us_equity"),
    );
    ([("content-type", "application/json")], HTTP_GET_ASSETS).into_response()
}

async fn handle_submit_order(headers: HeaderMap, body: Bytes) -> Response {
    assert_auth_headers(&headers);
    let request: serde_json::Value = serde_json::from_slice(&body).unwrap();
    assert_eq!(request["symbol"], "AAPL");
    assert_eq!(request["type"], "limit");
    assert_eq!(request["time_in_force"], "day");
    assert_eq!(request["qty"], "10");
    ([("content-type", "application/json")], HTTP_GET_ORDER).into_response()
}

#[tokio::test]
async fn client_get_account_sends_auth_headers_and_parses_response() {
    let base_url = spawn_server(Router::new().route("/v2/account", get(handle_account))).await;
    let client = authenticated_raw_client(&base_url);

    let account = client.get_account().await.unwrap();

    assert_eq!(account.id, "e6a5b6cd-1f27-4b3e-9c31-1a2c5d3f7c11");
    assert_eq!(account.status, "ACTIVE");
    assert_eq!(account.buying_power.as_deref(), Some("250000.00"));
}

#[tokio::test]
async fn client_request_instruments_parses_and_caches_equities() {
    let base_url = spawn_server(Router::new().route("/v2/assets", get(handle_assets))).await;
    let client = authenticated_client(&base_url);

    let instruments = client.request_instruments().await.unwrap();

    assert_eq!(instruments.len(), 2);
    assert_eq!(instruments[0].id().to_string(), "AAPL.ALPACA");

    let cached = client.get_instrument(&Ustr::from("AAPL")).unwrap();
    assert_eq!(cached.id(), instruments[0].id());
}

#[tokio::test]
async fn client_submit_order_posts_json_body() {
    let base_url = spawn_server(Router::new().route("/v2/orders", post(handle_submit_order))).await;
    let client = authenticated_client(&base_url);
    let params = PostOrderParamsBuilder::default()
        .symbol("AAPL")
        .qty("10")
        .side(AlpacaOrderSide::Buy)
        .order_type(AlpacaOrderType::Limit)
        .time_in_force(AlpacaTimeInForce::Day)
        .limit_price("189.05")
        .build()
        .unwrap();

    let order = client.submit_order(&params).await.unwrap();

    assert_eq!(order.id, "61e69015-8549-4bfd-b9c3-01e75843f47d");
    assert_eq!(order.status, AlpacaOrderStatus::Filled);
}

#[tokio::test]
async fn client_get_orders_sends_query_params() {
    async fn handler(Query(params): Query<std::collections::HashMap<String, String>>) -> Response {
        assert_eq!(params.get("status").map(String::as_str), Some("all"));
        assert_eq!(params.get("limit").map(String::as_str), Some("500"));
        ([("content-type", "application/json")], HTTP_GET_ORDERS).into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/orders", get(handler))).await;
    let client = authenticated_client(&base_url);
    let params = GetOrdersParamsBuilder::default()
        .status("all")
        .limit(500u32)
        .build()
        .unwrap();

    let orders = client.get_orders(&params).await.unwrap();

    assert_eq!(orders.len(), 2);
    assert_eq!(orders[1].status, AlpacaOrderStatus::PartiallyFilled);
}

#[tokio::test]
async fn client_cancel_order_accepts_204() {
    async fn handler(Path(order_id): Path<String>) -> Response {
        assert_eq!(order_id, "61e69015-8549-4bfd-b9c3-01e75843f47d");
        StatusCode::NO_CONTENT.into_response()
    }

    let base_url =
        spawn_server(Router::new().route("/v2/orders/{order_id}", delete(handler))).await;
    let client = authenticated_client(&base_url);

    client
        .cancel_order("61e69015-8549-4bfd-b9c3-01e75843f47d")
        .await
        .unwrap();
}

#[tokio::test]
async fn client_cancel_order_maps_422_to_venue_error() {
    async fn handler() -> Response {
        (
            StatusCode::UNPROCESSABLE_ENTITY,
            [("content-type", "application/json")],
            r#"{"code":42210000,"message":"order is not cancelable"}"#,
        )
            .into_response()
    }

    let base_url =
        spawn_server(Router::new().route("/v2/orders/{order_id}", delete(handler))).await;
    let client = authenticated_client(&base_url);

    let error = client.cancel_order("abc").await.unwrap_err();

    match error {
        AlpacaHttpError::Venue { code, message } => {
            assert_eq!(code, 42_210_000);
            assert!(message.contains("not cancelable"));
        }
        other => panic!("expected venue error, got {other:?}"),
    }
}

#[tokio::test]
async fn client_patch_order_returns_replacement() {
    async fn handler(Path(order_id): Path<String>, body: Bytes) -> Response {
        assert_eq!(order_id, "61e69015-8549-4bfd-b9c3-01e75843f47d");
        let request: serde_json::Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(request["qty"], "20");
        ([("content-type", "application/json")], HTTP_GET_ORDER).into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/orders/{order_id}", patch(handler))).await;
    let client = authenticated_client(&base_url);
    let params = PatchOrderParamsBuilder::default()
        .qty("20")
        .build()
        .unwrap();

    let order = client
        .patch_order("61e69015-8549-4bfd-b9c3-01e75843f47d", &params)
        .await
        .unwrap();

    assert_eq!(order.id, "61e69015-8549-4bfd-b9c3-01e75843f47d");
}

#[tokio::test]
async fn client_get_positions_parses_response() {
    async fn handler(headers: HeaderMap) -> Response {
        assert_auth_headers(&headers);
        ([("content-type", "application/json")], HTTP_GET_POSITIONS).into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/positions", get(handler))).await;
    let client = authenticated_client(&base_url);

    let positions = client.get_positions().await.unwrap();

    assert_eq!(positions.len(), 1);
    assert_eq!(positions[0].symbol, Ustr::from("AAPL"));
}

#[derive(Clone)]
struct PaginatedBarsState {
    calls: Arc<AtomicUsize>,
}

async fn handle_paginated_bars(
    State(state): State<PaginatedBarsState>,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Response {
    let call = state.calls.fetch_add(1, Ordering::SeqCst);
    if call == 0 {
        assert!(!params.contains_key("page_token"));
        ([("content-type", "application/json")], HTTP_GET_STOCK_BARS).into_response()
    } else {
        assert_eq!(
            params.get("page_token").map(String::as_str),
            Some("QUFQTHxNfDE3Mz"),
        );
        (
            [("content-type", "application/json")],
            HTTP_GET_STOCK_BARS_PAGE2,
        )
            .into_response()
    }
}

#[tokio::test]
async fn client_get_stock_bars_paginated_follows_page_tokens() {
    let state = PaginatedBarsState {
        calls: Arc::new(AtomicUsize::new(0)),
    };
    let base_url = spawn_server(
        Router::new()
            .route("/v2/stocks/bars", get(handle_paginated_bars))
            .with_state(state.clone()),
    )
    .await;
    let client = authenticated_client(&base_url);
    let params = GetStockBarsParamsBuilder::default()
        .symbols("AAPL")
        .timeframe("1Min")
        .build()
        .unwrap();

    let bars = client.get_stock_bars_paginated(&params).await.unwrap();

    assert_eq!(state.calls.load(Ordering::SeqCst), 2);
    assert_eq!(bars.get("AAPL").unwrap().len(), 3);
}

#[tokio::test]
async fn client_get_stock_trades_parses_response() {
    async fn handler() -> Response {
        (
            [("content-type", "application/json")],
            HTTP_GET_STOCK_TRADES,
        )
            .into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/stocks/trades", get(handler))).await;
    let client = authenticated_client(&base_url);
    let params = nautilus_alpaca::http::query::GetStockTradesParamsBuilder::default()
        .symbols("AAPL")
        .build()
        .unwrap();

    let trades = client.get_stock_trades_paginated(&params).await.unwrap();

    assert_eq!(trades.get("AAPL").unwrap().len(), 2);
}

#[tokio::test]
async fn client_get_stock_quotes_parses_response() {
    async fn handler() -> Response {
        (
            [("content-type", "application/json")],
            HTTP_GET_STOCK_QUOTES,
        )
            .into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/stocks/quotes", get(handler))).await;
    let client = authenticated_client(&base_url);
    let params = nautilus_alpaca::http::query::GetStockQuotesParamsBuilder::default()
        .symbols("AAPL")
        .build()
        .unwrap();

    let quotes = client.get_stock_quotes_paginated(&params).await.unwrap();

    assert_eq!(quotes.get("AAPL").unwrap().len(), 1);
}

#[derive(Clone)]
struct FlakyState {
    calls: Arc<AtomicUsize>,
}

async fn handle_flaky_account(State(state): State<FlakyState>, headers: HeaderMap) -> Response {
    assert_auth_headers(&headers);
    let call = state.calls.fetch_add(1, Ordering::SeqCst);
    if call == 0 {
        (StatusCode::INTERNAL_SERVER_ERROR, "boom").into_response()
    } else {
        ([("content-type", "application/json")], HTTP_GET_ACCOUNT).into_response()
    }
}

#[tokio::test]
async fn client_retries_5xx_and_succeeds() {
    let state = FlakyState {
        calls: Arc::new(AtomicUsize::new(0)),
    };
    let base_url = spawn_server(
        Router::new()
            .route("/v2/account", get(handle_flaky_account))
            .with_state(state.clone()),
    )
    .await;
    let client = authenticated_raw_client(&base_url);

    let account = client.get_account().await.unwrap();

    assert_eq!(state.calls.load(Ordering::SeqCst), 2);
    assert_eq!(account.status, "ACTIVE");
}

#[tokio::test]
async fn client_does_not_retry_venue_errors() {
    #[derive(Clone)]
    struct CountingState {
        calls: Arc<AtomicUsize>,
    }

    async fn handler(State(state): State<CountingState>) -> Response {
        state.calls.fetch_add(1, Ordering::SeqCst);
        (
            StatusCode::FORBIDDEN,
            [("content-type", "application/json")],
            HTTP_ERROR,
        )
            .into_response()
    }

    let state = CountingState {
        calls: Arc::new(AtomicUsize::new(0)),
    };
    let base_url = spawn_server(
        Router::new()
            .route("/v2/account", get(handler))
            .with_state(state.clone()),
    )
    .await;
    let client = authenticated_raw_client(&base_url);

    let error = client.get_account().await.unwrap_err();

    assert_eq!(state.calls.load(Ordering::SeqCst), 1);
    assert!(matches!(error, AlpacaHttpError::Venue { code, .. } if code == 42_210_000));
}

#[tokio::test]
async fn unauthenticated_client_can_fetch_market_data() {
    async fn handler(headers: HeaderMap) -> Response {
        assert!(headers.get("APCA-API-KEY-ID").is_none());
        (
            [("content-type", "application/json")],
            HTTP_GET_STOCK_BARS_PAGE2,
        )
            .into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/stocks/bars", get(handler))).await;
    let mut client = AlpacaHttpClient::new(
        AlpacaEnvironment::Paper,
        Some(base_url.clone()),
        Some(base_url),
        10,
        None,
    )
    .unwrap();
    // No-op: URLs already set via constructor; keep set_base_urls covered.
    let trading = client.base_url_trading().to_string();
    let data = client.base_url_data().to_string();
    client.set_base_urls(&trading, &data);

    let params = GetStockBarsParamsBuilder::default()
        .symbols("AAPL")
        .timeframe("1Min")
        .build()
        .unwrap();

    let response = client.get_stock_bars(&params).await.unwrap();

    assert_eq!(response.bars.get("AAPL").unwrap().len(), 1);
    assert_eq!(response.next_page_token, None);
}

#[tokio::test]
async fn client_get_assets_builder_defaults_produce_no_query() {
    async fn handler(Query(params): Query<std::collections::HashMap<String, String>>) -> Response {
        assert!(params.is_empty());
        ([("content-type", "application/json")], HTTP_GET_ASSETS).into_response()
    }

    let base_url = spawn_server(Router::new().route("/v2/assets", get(handler))).await;
    let client = authenticated_raw_client(&base_url);
    let params = GetAssetsParamsBuilder::default().build().unwrap();

    let assets = client.get_assets(&params).await.unwrap();

    assert_eq!(assets.len(), 2);
}
